### ISSUE TITLE
Subpar developer experience on Windows due to only binding to 127.0.0.1 by default

### ISSUE THREAD

**Issue by dmitri-gb** [fastify/fastify#3027]
_This issue is in fact a reiteration [of an older one](https://github.com/fastify/fastify/issues/1022). The previous issue was thought to be fixed, but turns out that's not the case._

## 🐛 Bug Report

Fastify binds its server socket to the `127.0.0.1` address by default. However, on a Windows system, `localhost` resolves to at least two addresses - `::1` and `127.0.0.1`, in that order. It seems that most modern HTTP clients first try to connect to `::1`, and once that fails, they try `127.0.0.1`. Ultimately this results in a noticable connection delay when trying out Fastify endpoints during development and using `localhost` as the hostname.

This issue was believed to be fixed by https://github.com/fastify/fastify/pull/1047. However, as I've already described [in a comment](https://github.com/fastify/fastify/pull/1047#issuecomment-823838719), it really wasn't. There was a misconception that passing `localhost` as the address to `Server.listen` will make it bind to `::1`, but that's not true.

## To Reproduce

The issue is fairly easy to reproduce on Windows. Create the most basic Fastify server (make sure to just use `.listen(port)` without any additional arguments) and make a request to it using `localhost` as the hostname. Notice that it takes a significant amount of time for the request to finish. Make the same request using `127.0.0.1` as the hostname and observe how the request is considerably faster. The exact results will of course depend on the client that is used to perform the requests. With curl, it's ~200ms, with [HTTPie](https://httpie.io/) it seems to be a whole **2 seconds**.

## Expected behavior

Ideally, the request should be equally fast for all possible loopback addresses, including `localhost`, `127.0.0.1` and perhaps `::1`.

There are multiple ways to improve this, though neither of them stands out as being a very good one.
1. Properly resolve `localhost` to the first entry returned by the OS and bind to that address. This way Fastify will bind to `::1` on Windows by default. It also means that connecting using `127.0.0.1` will not be possible. This was the intended fix of the original issue.
2. Bind to an IPv6 wildcard address (`::`), this is [the default behaviour of `Server.listen`](https://nodejs.org/dist/latest-v14.x/docs/api/net.html#net_server_listen_port_host_backlog_callback:~:text=If%20host%20is%20omitted) when not passing it any address at all. This is also a special case in that it will additionally bind to the IPv4 wildcard address (a so called dual-stack socket). This way all three desired loopback hosts will connect without delay. The downside, of course, is that the socket is open to everyone, and remote machines can also connect to it.
3. Create multiple sockets, one for each address that `localhost` resolves to. This is likely undesireable, as it significantly increases the complexity of the code.

It is probably also OK to just leave things as they are, and accept the fact that connecting using `localhost` on Windows will result in a delay if the bind address hasn't been specified explicitly.

In any case, [this comment in server.js](https://github.com/fastify/fastify/blob/2af8a3ceae2e498a664e9865f31920bd987e8ca2/lib/server.js#L63-L65) is misleading and should at the very least be removed.

## Your Environment

- *node version*: 14.16.1
- *fastify version*: 3.14.2
- *os*: Windows

**Comment by simoneb** [comment]
I've been using Fastify on Windows for a long time and I never came across this issue. I would like to understand if there's anything specific to your setup that would cause this. Any thoughts?

**Comment by zekth** [comment]
I personnaly observe 2ms on `127.0.0.1` and 230ms on `localhost` which is expected as we skip the DNS resolution. Using curl here are the results:
```
$ curl -s http://localhost:3000/health -w " %{time_total}s"
OK 0,234610s
$ curl -s http://127.0.0.1:3000/health -w " %{time_total}s"
OK 0,001137s
```

**Comment by simoneb** [comment]
@zekth is that the first request or any request? Can you share the code for that so we're using the same baseline for comparison?

**Comment by zekth** [comment]
I got overall the same time between first and any request. 
Here is the code:
```js
const fastify = require('fastify')({
  logger: true
})
fastify.get('/health', (request, reply) => {
  reply.send('OK')
})
fastify.listen(3000, (err, address) => {
  if (err) throw err
})
```
node v14.15.1
fastify 3.15.0

**Comment by simoneb** [comment]
Here's what I have:

```
$ curl -s http://localhost:3000/health -w " %{time_total}s"
OK 0.213064s
$ curl -s http://127.0.0.1:3000/health -w " %{time_total}s"
OK 0.001937s
```

Node v14.15.4
Fastify 3.15.0

**Comment by dmitri-gb** [comment]
Yes, those extra 200 ms that you both observe are in fact the manifestation of this very issue. If you modify your server to bind to the IPv6 loopback address instead (`::1`), then you'll see that making a request to `localhost` will be down to just a few ms.

The severity of this clearly depends on the client, and with curl it is less of an issue (not even noticable to many). But other clients may behave differently. Someone claimed that [Postman takes an extra 500 ms](https://github.com/fastify/fastify/issues/1022#issuecomment-404785277). While HTTPie takes an extra 2 seconds (though this is certainly something they could improve on). There exists, in fact, an IETF RFC that is meant to address this very issue, and it recommends that the IPv4 address be tried with a [150-250 ms delay](https://www.ietf.org/rfc/rfc6555.txt#:~:text=150%2D250%20ms%20apart).

The main goal of me creating this issue was to point out that the Fastify team once _intended_ to change the default bind address to `::1`, but that change didn't actually have any effect (and resulted in a misleading comment in the code). So by all means, it is fine to leave the binding scheme the way it is 🙂

**Comment by simoneb** [comment]
@dmitri-gb actually this is indeed an interesting issue that working on Windows I never noticed and it may indeed in fact be the (main?) reason why ci is [considerably slower](https://github.com/fastify/fastify/runs/2413966841?check_suite_focus=true) on Windows than it is [on Linux](https://github.com/fastify/fastify/runs/2413966816?check_suite_focus=true).

Apart from acknowledging that this issue exists, do you think there is actually a way to address it?

**Comment by jsumners** [comment]
I really don't know what to do here. `server.listen` does not allow for passing of options through to `dns.lookup`. It also does not allow for an array of bind paths to be specified. The only way to make this work as desired is to use https://github.com/nodejs/node/pull/38099 once it lands. But that only covers some release of v16 and forward.

As I originally noted, https://github.com/fastify/fastify/pull/1047#issuecomment-406441487, I think clarifying the documentation is the only thing we can do and retain our secure first implementation.

**Comment by jsumners** [comment]
> it may indeed in fact be the (main?) reason why ci is [considerably slower](https://github.com/fastify/fastify/runs/2413966841?check_suite_focus=true) on Windows than it is [on Linux](https://github.com/fastify/fastify/runs/2413966816?check_suite_focus=true).

It's certainly a contributing factor, but the the slowness of Windows CI is really just down to it being Windows. It's more expensive to spin up a Windows container than a native Linux container.

**Comment by dmitri-gb** [comment]
> `server.listen` does not allow for passing of options through to `dns.lookup`.

If you wanted to go the route of binding to `::1` on Windows, then Fastify could perform the lookup itself by invoking `dns.lookup` directly.

**Comment by mcollina** [comment]
> > `server.listen` does not allow for passing of options through to `dns.lookup`.
> 
> 
> 
> If you wanted to go the route of binding to `::1` on Windows, then Fastify could perform the lookup itself by invoking `dns.lookup` directly.
> 
> 

Could you send a PR? I'm not exactly sure how that would work.

**Comment by dmitri-gb** [comment]
@mcollina, something like this: https://github.com/dmitri-gb/fastify/commit/34b095090e7400e76607445cfc587335b8d6ad1d

I can submit the change as a PR if you wish, but many tests are failing on Windows due to the fact that `simple-get` still tries to connect to `127.0.0.1`. I'm afraid I don't have time to flesh out the PR completely though.

**Comment by rhadamez** [comment]
Well, I'm testing fastify for 2 days, and I just faced this problem, thanks to someone that show me this issue, I was frustrated by the performance of 250ms to make a simple post. Changing the request origin to 127.0.0.1 solved the problem.

**Comment by climba03003** [comment]
> 
> 
> @mcollina, something like this: [dmitri-gb@34b0950](https://github.com/dmitri-gb/fastify/commit/34b095090e7400e76607445cfc587335b8d6ad1d)
> 
> I can submit the change as a PR if you wish, but many tests are failing on Windows due to the fact that `simple-get` still tries to connect to `127.0.0.1`. I'm afraid I don't have time to flesh out the PR completely though.

If it will break the test, it is creating problem instead of fixing the issue. Because not all the people is using `localhost` to connect the server when they are dev. Some of them will be using `127.0.0.1`, that means if we land your fix, it will be breaking the ppl using `127.0.0.1` in Windows.

**Comment by dmitri-gb** [comment]
> that means if we land your fix, it will be breaking the ppl using `127.0.0.1` in Windows.

Exactly. I already pointed this out in the original post and said that none of the three potential improvements stands out as being particularly good.

> unless `http` allow to pass in binding address as Array

Binding to multiple non-wildcard addresses requires creating multiple sockets. You would have to do `http.createServer(...)` twice with the same handler, however it also requires housekeeping and error handling of both sockets and that is simply a lot of code with too little benefit.

I agree that simply mentioning and clarifying this issue in the documentation is a perfectly fine resolution (along with the removal of [this comment](https://github.com/fastify/fastify/blob/2af8a3ceae2e498a664e9865f31920bd987e8ca2/lib/server.js#L63-L65)).

**Comment by mcollina** [comment]
> Binding to multiple non-wildcard addresses requires creating multiple sockets. You would have to do http.createServer(...) twice with the same handler, however it also requires housekeeping and error handling of both sockets and that is simply a lot of code with too little benefit.

Would this be possible at the lower levels, i.e. C, or it is a unix restriction?

**Comment by mcollina** [comment]
Would you like to send a PR to fix the docs and comment?

**Comment by Fdawgs** [comment]
Closed by #3606
